File: aula1/credit.c
#include<stdio.h>
#include<cs50.h>
#include<math.h>

int checksum(long card_number);

int main(void)
{
    long card_number = get_long("Card#\n");

    if (checksum(card_number) % 10 == 0)
    {
        //printf("OK\n");
        //printf("%i\n",(int)(card_number / pow(10, 14)));
        if ((int)(card_number / pow(10, 13)) == 34 || (int)(card_number / pow(10, 13)) == 37)
        {
            printf("AMEX\n");
        }
        else if ((int)(card_number / pow(10, 14)) > 50 && /*50 <*/ (int)(card_number / pow(10, 14)) < 56)
        {
            printf("MASTERCARD\n");
        }
        else if (pow(10, 12) < card_number < pow(10, 13) && (int)(card_number / pow(10, 12)) == 4)
        {
            printf("VISA\n");
        }
        else if (pow(10, 15) < card_number < pow(10, 16) && (int)(card_number / pow(10, 15)) == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int checksum(long card_number)
{
    long number = round(card_number / 10); //Assign card_number to a test variable
    int sum = 0;

    do
    {
        if (number % 10 * 2 < 9)
        {
            sum += (number % 10) * 2;
        }
        else
        {
            sum += ((number % 10) * 2) % 10 + 1;
        }
        //printf("%i\n", sum);
        number = round(number / 100);
    }
    while (number > 0);

    //printf("%i\n", sum);

    number = card_number; //Reassign card_number to a test variable
    do
    {
        sum += (number % 10);
        number = round(number / 100);
    }
    while (number > 0);

    //printf("%i\n", sum);
    return sum;
}

File: cs50/aula6/credit.py
from cs50 import get_int
from cs50 import get_string

def main():
    card_int = get_int("Number: ")
    card = str(card_int)
    number = []
    valid_len = [13, 15, 16]
    if len(card) not in valid_len:
        print("INVALID")
        return 0

    for i in range(0, len(card)):
        number.append(int(card[i]))

    if (valid(number, card)):
        # print(f"{len(card)} & {card[0]} & {card[1]}")
        # print(f"{number[0]} & {number[1]}")

        if len(card) == 15 and number[0] == 3 and (number[1] == 4 or number[1] == 7): # in [4, 7]:
            print("AMAAMEX")

        if len(card) == 16 and number[0] == 5 and (number[1] > 0 or number[1] < 6):# in range(1, 5):
            print("MASTERCARD")

        if len(card) in [13, 16] and number[0] == 4:
            print("VISA")
        else:
            print("INVALID")

def valid(number, card):

    sum = 0
    for i in range(len(card) - 2, -1, -2):
        sum += (((number[i] * 2) // 10) + (number[i] * 2) % 10)
        # print(f"{number[i]} * 2 = {sum}")

    for i in range(len(card) - 1, -1, -2):
        # print(number[i])
        sum += number[i]
        # print(f"{number[i]} {sum}")

    if sum % 10 != 0:
        print("INVALID")
        return False
    else:
        return True

main()


